src/main/java/org/folio/services/locationunit/InstitutionService.java

```diff
@@ -0,0 +1,117 @@
+import static org.folio.rest.impl.LocationUnitApi.URL_PREFIX;
+import static org.folio.rest.tools.utils.ValidationHelper.createValidationErrorMessage;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.sqlclient.Row;
+import io.vertx.sqlclient.RowSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import javax.ws.rs.core.Response;
+import lombok.extern.log4j.Log4j2;
+import org.folio.persist.InstitutionRepository;
+import org.folio.rest.jaxrs.model.Locinst;
+import org.folio.rest.jaxrs.model.Locinsts;
+import org.folio.rest.jaxrs.resource.LocationUnits.DeleteLocationUnitsInstitutionsByIdResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.DeleteLocationUnitsInstitutionsResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.GetLocationUnitsInstitutionsByIdResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.GetLocationUnitsInstitutionsResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.PostLocationUnitsInstitutionsResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.PutLocationUnitsInstitutionsByIdResponse;
+import org.folio.rest.persist.PgUtil;
+import org.folio.services.domainevent.InstitutionDomainEventPublisher;
+
+@Log4j2
+public class InstitutionService {
+
+  public static final String INSTITUTION_TABLE = "locinstitution";
+  private static final String MSG_ID_NOT_MATCH =
+    "Illegal operation: Institution ID cannot be changed";
+
+  private final Context vertxContext;
+  private final Map<String, String> okapiHeaders;
+  private final InstitutionRepository repository;
+  private final InstitutionDomainEventPublisher domainEventService;
+
+  public InstitutionService(Context vertxContext,
+                            Map<String, String> okapiHeaders) {
+    this.vertxContext = vertxContext;
+    this.okapiHeaders = okapiHeaders;
+    this.repository = new InstitutionRepository(vertxContext, okapiHeaders);
+    this.domainEventService =
+      new InstitutionDomainEventPublisher(vertxContext, okapiHeaders);
+  }
+
+  public Future<Response> getByQuery(String cql, int offset, int limit) {
+    return PgUtil.get(INSTITUTION_TABLE, Locinst.class, Locinsts.class, cql,
+      offset, limit, okapiHeaders, vertxContext,
+      GetLocationUnitsInstitutionsResponse.class);
+  }
+
+  public Future<Response> getById(String id) {
+    return PgUtil.getById(INSTITUTION_TABLE, Locinst.class, id, okapiHeaders,
+      vertxContext, GetLocationUnitsInstitutionsByIdResponse.class);
+  }
+
+  public Future<Response> create(Locinst institution) {
+    return PgUtil.post(INSTITUTION_TABLE, institution, okapiHeaders,
+        vertxContext, PostLocationUnitsInstitutionsResponse.class)
+      .onSuccess(response -> {
+        domainEventService.publishCreated().handle(response);
+
+        var headers = PostLocationUnitsInstitutionsResponse.headersFor201()
+          .withLocation(URL_PREFIX + response);
```

Comment:
do not know who use this locations header, as for me it is ok to have the same approach as in similar PR to make code simpler and cleaner:
https://github.com/folio-org/mod-inventory-storage/pull/1029/files#diff-a504c82181f298bf1db3c82e991febfb96b01ed3ed5d5295ea46050e1a2b2fcfR54
But let's wait what @psmagin will suggest

Reply:
I'm okay with not having it